### PR TITLE
437 implement #if_#then_#else hook

### LINKED ISSUE #437: Implement the `KEQUAL.ite` hook tp handle `#if` on Booster
In `evm-semantics`, a new `nogas` mode is being implemented, which introduces the `<use-gas>` boolean cell in the configuration. Depending of the value of that cell, which will always be concrete, some rules either compute or do not compute gas usage. To prevent duplicating these rules for the yes-gas and no-gas cases, the computation involves a `#if` expression using the concrete switch above. The `#if` is not evaluated by Booster, since we do not implement `KEQUAL.ite`.

Since we have already implemented `KEQUAL.eq` and `KEQUAL.non-eq`, we may as well implement `KEQUAL.ite`.

### PR DESCRIPTION
**Pull request by jberthold**
* Evaluation now proceeds bottom-up (but still discovers work top-down)
* New module `Booster.Builtin` as a home for all hook implementations
* If a hook implementation is provided, it will be tried before function and simplification equations
  * A hook is allowed to return no result, in which case function and simplification equations will be tried
  * Sort errors or arity errors (currently) abort the evaluation (should be caught during internalisation)

FIxes #437

### CODE REVIEW

File: test/rpc-integration/test-simplify/response-if-then-else-sort-error.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+        "data": {
+            "context": [
+                "symbol or alias 'inj' (<unknown location>)",
+                "symbol or alias 'Lbl'Hash'if'UndsHash'then'UndsHash'else'UndsHash'fi'Unds'K-EQUAL-SYNTAX'Unds'Sort'Unds'Bool'Unds'Sort'Unds'Sort' (<unknown location>)",
+                "(resources/simplify.kore 86:139, <unknown location>)"
+            ],
+            "error": "Expecting sort SortBool{} but got SortInt{}."
+        },
+        "message": "Could not verify pattern"
+    }
```
Comment (jberthold) [inline test/rpc-integration/test-simplify/response-if-then-else-sort-error.json:15]: NB This error is from `kore-rpc` and produced after hitting a failure in `booster`, I opened #443 to address the issue.

File: library/Booster/Builtin.hs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                [ "Different sorts in alternatives:"
+                , pretty thenVal
+                , pretty elseVal
+                ]
+        case cond of
+            TrueBool -> pure $ Just thenVal
+            FalseBool -> pure $ Just elseVal
+            _other -> pure Nothing
+    | otherwise =
+        throwE . renderText $ "if-then-else: wrong arity " <> pretty (length args)
```
Comment (geo2a) [inline library/Booster/Builtin.hs:63]:
```suggestion
        throwE . renderText $ "KEQUAL.ite: wrong arity " <> pretty (length args)
```

**Comment by jberthold** [comment]
KEVM tests show no change - even an empty file when comparing with default

| Test                                     | 437-if-then-else-hook time | master-4a693bc time | (437-if-then-else-hook/master-4a693bc) time 
|------------------------------------------|----------------------------|---------------------|---------------------------------------------
| erc20/ds/transferFrom-success-2-spec.k   | 95.83                      | 98.5                | 0.9728934010152284                          
| erc20/ds/transfer-failure-2-a-spec.k     | 69.37                      | 70.64               | 0.982021517553794                           
| benchmarks/storagevar02-overflow-spec.k  | 88.71                      | 90.19               | 0.9835901984698968                          
| mcd/vat-muluu-pass-spec.k                | 107.82                     | 109.58              | 0.9839386749406825                          
| bihu/forwardToHotWallet-failure-2-spec.k | 132.66                     | 134.74              | 0.9845628618079263                          
| erc20/hkg/transfer-success-2-spec.k      | 51.47                      | 52.27               | 0.9846948536445379                          
| benchmarks/encode-keccak00-spec.k        | 53.31                      | 54.06               | 0.9861265260821309                          
| bihu/functional-spec.k                   | 116.47                     | 118.0               | 0.9870338983050847                          
| benchmarks/requires01-a0gt0-spec.k       | 51.55                      | 52.19               | 0.9877371143897298                          
| benchmarks/storagevar00-spec.k           | 64.63                      | 65.42               | 0.987924182207276                           
| erc20/ds/transferFrom-failure-1-c-spec.k | 123.77                     | 125.15              | 0.9889732321214542                          
| benchmarks/staticloop00-a0lt10-spec.k    | 53.0                       | 53.57               | 0.9893597162591002                          
| benchmarks/encodepacked-keccak01-spec.k  | 48.82                      | 49.32               | 0.9898621248986212                          
| erc20/ds/transferFrom-failure-1-b-spec.k | 153.09                     | 154.63              | 0.9900407424173835                          
| erc20/hkg/transferFrom-failure-1-spec.k  | 104.87                     | 103.83              | 1.0100163729172686                          
| erc20/ds/transfer-success-1-spec.k       | 73.83                      | 73.08               | 1.0102627257799672                          
| mcd/flipper-bids-pass-rough-spec.k       | 107.79                     | 106.63              | 1.0108787395667262                          
| bihu/forwardToHotWallet-failure-3-spec.k | 102.38                     | 101.15              | 1.0121601581809194                          
| erc20/ds/transferFrom-failure-1-d-spec.k | 103.59                     | 101.97              | 1.0158870255957635                          
| TOTAL                                    | 1702.9599999999998         | 1714.9199999999998  | 0.9930259137452476  

kontrol tests got borken by latest merges making booster default (the `--use-booster` is rejected), rerunning now

**Comment by jberthold** [comment]
| Test                                              | 437-if-then-else-hook time | master-4a693bc time | (437-if-then-else-hook/master-4a693bc) time 
|---------------------------------------------------|----------------------------|---------------------|---------------------------------------------
| BlockParamsTest.testBlockNumber()                 | 31.43                      | 33.67               | 0.9334719334719335                          
| AssumeTest.test_assume_true(uint256,uint256)      | 29.82                      | 31.24               | 0.9545454545454546                          
| AccountParamsTest.test_Nonce_NonExistentAddress() | 43.57                      | 45.53               | 0.9569514605754448                          
| AddrTest.test_notBuiltinAddress_symbolic(address) | 46.2                       | 48.16               | 0.9593023255813955                          
| ExpectRevertTest.testFail_expectRevert_empty()    | 30.68                      | 31.95               | 0.9602503912363067                          
| GasTest.testSetGas()                              | 25.69                      | 24.48               | 1.0494281045751634                          
| TOTAL                                             | 207.39                     | 215.02999999999997  | 0.9644700739431707

**Comment by jberthold** [comment]
Open question in this context: (Why) Do we need to have the `K` sort on the rules for `#if` in `domains.md`? 
https://github.com/runtimeverification/k/blob/master/k-distribution/include/kframework/builtin/domains.md?plain=1#L2304-L2305

**Comment by ehildenb** [comment]
I'm pretty sure the old backend actually applies hooks _after_ user-supplied simplifications. This is because the use may want to do something more structured than what the hook does, by recognizing a specific pattern and rewriting it to something else.

It may be good to follow the same behavior, but good to check first.

**Comment by ehildenb** [comment]
Bascially, if the user supplies rules for a particular scenario, we should use those rules. This is under the general principle of "the tool should be responsive to user input, rather than relying on magic".

**Comment by jberthold** [comment]
> I'm pretty sure the old backend actually applies hooks _after_ user-supplied simplifications. This is because the use may want to do something more structured than what the hook does, by recognizing a specific pattern and rewriting it to something else.

At the moment, the old backend [applies hooks before user-supplied equations (simplifications and equations)](https://github.com/runtimeverification/haskell-backend/blob/master/kore/src/Kore/Rewrite/Function/Evaluator.hs#L316-L321). 

That said, the idea to follow the user's guidelines first before using a hook is definitely worth considering. We just need to make sure there are no hidden issues from rules in the K prelude (`domains.md` ) not being coherent to the hook behaviour.

**Review by geo2a — APPROVED** [review]

File: library/Booster/Builtin.hs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                [ "Different sorts in alternatives:"
+                , pretty thenVal
+                , pretty elseVal
+                ]
+        case cond of
+            TrueBool -> pure $ Just thenVal
+            FalseBool -> pure $ Just elseVal
+            _other -> pure Nothing
+    | otherwise =
+        throwE . renderText $ "KEQUAL.ite: wrong arity " <> pretty (length args)
```
Comment (goodlyrottenapple) [inline library/Booster/Builtin.hs:63]: This is nice! Could we hook `==K` in the same way?

**Review by goodlyrottenapple — APPROVED** [review]

File: library/Booster/Builtin.hs (hunk 3 above)
Comment (jberthold) [reply library/Booster/Builtin.hs:63]: Yes, I was going to do that, but not in this same PR.

**Pull request merged**